File: shpark/employeeImportance.java
import java.util.*;

class Employee {
	// It's the unique id of each node;
	// unique id of this employee
	public int id;
	// the importance value of this employee
	public int importance;
	// the id of direct subordinates
	public List<Integer> subordinates;
};

public class employeeImportance {
	Boolean[] visited;

	public int getImportance(List<Employee> employees, int id) {
		int employeeCnt = employees.size();
		visited = new Boolean[2000];
		int[] idxarr = new int[2000];
		for(int i=0; i<2000;i++){
			idxarr[i]=-1;
			visited[i]=false;
		}
		for (int i = 0; i < employeeCnt; i++) {
			idxarr[employees.get(i).id - 1] = i;
		}
		
		return helper(id, idxarr, employees);
	}

	public int helper(int id, int[] idxarr, List<Employee> employees) {
		Employee emp = employees.get(idxarr[id - 1]);
		int totalImportance = 0;
		if (!visited[idxarr[id - 1]]) {
			visited[idxarr[id - 1]] = true;
			totalImportance += emp.importance;
		}
		for (Integer sId : emp.subordinates) {
			totalImportance += helper(sId, idxarr, employees);
		}
		return totalImportance;
	}
}

File: byssup/dfs/Employee Importance.py
"""
# Employee info
class Employee(object):
    def __init__(self, id, importance, subordinates):
        # It's the unique id of each node.
        # unique id of this employee
        self.id = id
        # the importance value of this employee
        self.importance = importance
        # the id of direct subordinates
        self.subordinates = subordinates
"""


class Solution(object):
    def getImportance(self, employees, id):
        """
        :type employees: Employee
        :type id: int
        :rtype: int
        """
        temp = [0] * 2001
        for emp in employees:
            temp[emp.id] = (emp.importance, emp.subordinates)
        return self.get_imp(temp, id)

    def get_imp(self, employess, id):
        if not employess[id][1]:
            return employess[id][0]

        imp = employess[id][0]
        for sub_id in employess[id][1]:
            imp += self.get_imp(employess, sub_id)
        return imp